### PR TITLE
Add warning on old style include statement

### LINKED ISSUE #2568: include does not work any more
# syslog-ng

## Version of syslog-ng

syslog-ng compiled from git master today

## Platform

multiple openSUSE versions

# Issue

## Failure

Sometime after 3.19.1 release handling of include broke, the syslog-ng.conf used for the past couple of years does not work any more:

```
[...]
[2019-02-25T18:24:49.071499] Module loaded and initialized successfully; module='csvparser'
[2019-02-25T18:24:49.071548] Finishing include; content='block parser sudo-parser() at /usr/share/syslog-ng/include/scl/sudo/sudo.conf:23', depth='3'
[2019-02-25T18:24:49.071575] Finishing include; content='parser generator app-parser', depth='2'
[2019-02-25T18:24:49.071597] Finishing include; content='source generator system', depth='1'
Error parsing config, syntax error, unexpected KW_INCLUDE, expecting $end in /etc/syslog-ng/syslog-ng.conf:43:1-43:8:
38      	# uncomment to process log messages from network:
39      	#
40      	#udp(ip("0.0.0.0") port(2000));
41      };
42      
43----> include "/run/syslog-ng/additional-log-sockets.conf";
43----> ^^^^^^^
44      
45      #
46      # Filter definitions
47      #
48      filter f_iptables   { facility(kern) and message("IN=") and message("OUT="); };


syslog-ng documentation: https://www.balabit.com/support/documentation?product=syslog-ng-ose
contact: https://lists.balabit.hu/mailman/listinfo/syslog-ng
```

Downgrading to 3.19.1 resolves the problem.

## Configuration

```
@version:3.19
@include "scl.conf"

#
# /etc/syslog-ng/syslog-ng.conf
#
# File format description can be found in syslog-ng.conf(5)
# and in /usr/share/doc/packages/syslog-ng/syslog-ng.txt.
#
# NOTE: The SuSEconfig script and its syslog-ng.conf.in
#       configuration template aren't used any more.
#
#       Feel free to edit this file directly.
#

#
# Global options.
#
options { chain_hostnames(off); flush_lines(0); perm(0640); stats_freq(3600); threaded(yes); };

#
# 'src' is our main source definition. you can add
# more sources driver definitions to it, or define
# your own sources, i.e.:
#
#source my_src { .... };
#
source src {
        #
        # use system() for local logs
        #
        system();
        #
        # syslog-ng's internal messages
        #
        internal();
        #
        # uncomment to process log messages from network:
        #
        #udp(ip("0.0.0.0") port(2000));
};

include "/run/syslog-ng/additional-log-sockets.conf";

#
# Filter definitions
#
filter f_iptables   { facility(kern) and message("IN=") and message("OUT="); };

filter f_console    { level(warn) and facility(kern) and not filter(f_iptables)
                      or level(err) and not facility(authpriv); };
[...]
```

### PR DESCRIPTION
**Pull request by bazsi**
This patch attempts to resolve #2568 by patching the input stream before the lexer/grammar processes it. This will also open up further avenues to patch the configuration file so that old constructs can be removed from the grammar.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/cfg-lex.l [hunk 1]
```diff
@@ -71,13 +71,64 @@ yy_input_run_backtick_substitution(CfgLexer *self, gchar *buf, gsize buf_size, g
...
+    {
+      msg_error("Error performing the `include' patching to @include, the target buffer is too small",
+                evt_tag_int("buf_size", buf_size),
+                evt_tag_str("filename", cur_lloc->level->name),
+                evt_tag_printf("line", "%d:%d", cur_lloc->first_line, cur_lloc->first_column));
+      return FALSE;
+    }
+
+  memmove(buf + 1, buf, *len);
+  buf[0] = '@';
```
[the remark this answers is not in the document]
Comment (bazsi) [reply lib/cfg-lex.l]: Thanks for pointing this out.

**Comment by Kokan** [comment]
Do we want to support the pragma less include in the long term ?
Because if we don't want to, I do not see a reason to do backward compatibility when the `3.20.1` has been released before it.
The users are already effected, unless we are planning to do a `3.20.2` release to address this.

**Comment by czanik** [comment]
The worst hit is the openSUSE / SLES community, where tens of thousands of users have in their default syslog-ng.conf:
```
include "/run/syslog-ng/additional-log-sockets.conf";
```
Anybody who installed syslog-ng in the past 5-6 years, either from an official distro package or from my unofficial packages.
If they ever modified /etc/syslog-ng/syslog-ng.conf then RPM leaves it alone and the new config is saved to a different file. What it means in practice, that syslog-ng does not start after an upgrade.

@Kokan openSUSE Tumbleweed and the upcoming openSUSE Leap 15.1 both have syslog-ng 3.19.1. My own unofficial syslog-ng package built from 3.20.1 sources is now patched with this PR: https://build.opensuse.org/package/show/home:czanik:syslog-ng320/syslog-ng So a separate 3.20.2 is not necessary from the openSUSE / SLES point of view, where this change can cause a problem in large scale.

@bazsi Do I understand well, that the original intent was, that with old configuration versions users should get a warning but be able to use unmodified config when using ```include``` without a ```@``` sign? There is one more problem: the ```;``` at the end:
```
linux-6chp:~ # syslog-ng -s
[2019-03-05T11:17:52.095617] WARNING: Configuration file format is too old, syslog-ng is running in compatibility mode. Please update it to use the syslog-ng 3.20 format at your time of convenience. To upgrade the configuration, please review the warnings about incompatible changes printed by syslog-ng, and once completed change the @version header at the top of the configuration file.;
[2019-03-05T11:17:52.109831] WARNING: the `include' keyword based inclusion syntax (e.g. the one without a '@' character in front) has been removed in syslog-ng 3.20. Please prepend a '@' character in front of your include statement while updating your configuration to match the new version; filename='/etc/syslog-ng/syslog-ng.conf', line='43:1'
Error parsing pragma, syntax error, unexpected ';', expecting LL_EOL in /etc/syslog-ng/syslog-ng.conf:43:54-43:55:
38      	# uncomment to process log messages from network:
39      	#
40      	#udp(ip("0.0.0.0") port(514));
41      };
42      
43----> include "/run/syslog-ng/additional-log-sockets.conf";
43---->                                                      ^
44      
45      #
46      # Filter definitions
47      #
48      filter f_iptables   { facility(kern) and message("IN=") and message("OUT="); };

Error parsing config, syntax error, unexpected LL_ERROR, expecting $end in /etc/syslog-ng/syslog-ng.conf:43:1-43:55:
38      	# uncomment to process log messages from network:
39      	#
40      	#udp(ip("0.0.0.0") port(514));
41      };
42      
43----> include "/run/syslog-ng/additional-log-sockets.conf";
43----> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
44      
45      #
46      # Filter definitions
47      #
48      filter f_iptables   { facility(kern) and message("IN=") and message("OUT="); };


syslog-ng documentation: https://www.balabit.com/support/documentation?product=syslog-ng-ose
contact: https://lists.balabit.hu/mailman/listinfo/syslog-ng
```
Unfortunately it still fails with an error.

**Comment by bazsi** [comment]
uggh, the final ';' is not handled yet, unfortunately, I've missed that
piece but should be trivial to fix the same wy.

**Pull request merged**